### PR TITLE
Verify ceph-ansible log file is created and contains the logs of a pa…

### PR DESCRIPTION
**Pull request by adityaramteke**
This PR is to verify the ceph-ansible log file is created and contains the logs of a particular playbook!

The following steps are needed to perform the test for CEPH-83571467:

- Set the desired value to the log_path at /usr/share/ceph-ansible/ansible.cfg
- Run the deployment playbooks (example - site.yml, purge-cluster.yml, rolling-update.yml)
- Log file should have been created/appended with the respective playbooks initiated and has the appropriate contents.

Run result: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-443W6Z/ceph_ansible_0.log

Signed-off-by: Aditya Ramteke <aramteke@redhat.com>

### CODE REVIEW

**Review by Ameenasuhani — COMMENTED** [review]
The requirement was to capture the logs from  execution of all playbooks. The workflow looks like only site.yml is being considered? Can we re discuss on the requirement?

https://polarion.engineering.redhat.com/polarion/#/project/CEPH/workitem?id=CEPH-83572701

**Comment by adityaramteke** [comment]
> The requirement was to capture the logs from execution of all playbooks. The workflow looks like only site.yml is being considered, can we re discuss on the requirement?
> 
> https://polarion.engineering.redhat.com/polarion/#/project/CEPH/workitem?id=CEPH-83572701

Thanks, @Ameenasuhani for your comment and offline discussion on this! I'll re-design this PR and send it for the review!

File: suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml [hunk 1]
```diff
@@ -234,3 +234,22 @@ tests:
       polarion-id: CEPH-83573277
       module: purge_dashboard.py
       desc: Purges ceph dashboard
+
+  - test:
+      name: Ansible.log update
```
Comment (sunilangadi2) [inline suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml:239]: based on the test name CI log file will be created so let's not provide any special char in test name remove . just keep it simple as ` anisble log collection`?

**Review by sunilangadi2 — CHANGES_REQUESTED** [review]

File: suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml (hunk 1 above)
Comment (adityaramteke) [reply suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml]: > based on the test name CI log file will be created so let's not provide any special char in test name remove . just keep it simple as ` anisble log collection`?
@sunilangadi2 can we say `Ceph ansible log validation` instead?

File: suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml [hunk 2]
```diff
@@ -234,3 +234,43 @@ tests:
       polarion-id: CEPH-83573277
       module: purge_dashboard.py
       desc: Purges ceph dashboard
+
+  - test:
+      name: Ceph ansible log validation
+      polarion-id: CEPH-83572701
+      module: test_ansible.py
+      config:
```
Comment (sunilangadi2) [inline suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml:242]: @adityaramteke why you need all these config here?

Comment (adityaramteke) [reply suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml:242]: @sunilangadi2 sure, we can do that!

Comment (adityaramteke) [reply suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml]: @sunilangadi2 the requested changes are updated! :-)

File: suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml [hunk 3]
```diff
@@ -26,6 +26,7 @@ tests:
       module: test_ansible.py
       config:
         is_mixed_lvm_configs: True
+        custom_log: true # CEPH-83572701
```
Comment (psathyan) [inline suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml:29]: Please modify the `polarion-id` key instead of adding a comment.

File: tests/ceph_installer/test_ansible.py [hunk 4]
```diff
@@ -119,7 +119,16 @@ def run(ceph_cluster, **kw):
...
-    LOG.info("Ceph ansible version " + ceph_installer.get_installed_ceph_versions())
+    # Check if ansilbe.log file has any logs initially
+    if custom_log:
+        ansible_log_file, _ = ceph_installer.exec_command(
+            cmd="cat {}/ansible.cfg | grep log_path | cut -d \" \" -f3 | tr -d '\n\r'".format(
+                ansible_dir
+            )
+        )
+        file_size_before, _ = ceph_installer.exec_command(
+            cmd='ls -l "{}" | cut -d " " -f5 | tr -d \'\n\r\''.format(ansible_log_file)
```
Comment (psathyan) [inline tests/ceph_installer/test_ansible.py:130]: Could we use the command `du`?

File: tests/ceph_installer/test_ansible.py [hunk 5]
```diff
@@ -119,7 +119,16 @@ def run(ceph_cluster, **kw):
     if test_data.get("luns_setting", None) and test_data.get("initiator_setting", None):
         ceph_installer.add_iscsi_settings(test_data)
 
-    LOG.info("Ceph ansible version " + ceph_installer.get_installed_ceph_versions())
+    # Check if ansilbe.log file has any logs initially
+    if custom_log:
+        ansible_log_file, _ = ceph_installer.exec_command(
+            cmd="cat {}/ansible.cfg | grep log_path | cut -d \" \" -f3 | tr -d '\n\r'".format(
```
Comment (psathyan) [inline tests/ceph_installer/test_ansible.py:125]: It is recommended to use format strings i.e. `cmd=f"cat {ansible_dir}/ansible.cfg | ... "`

File: tests/ceph_installer/test_ansible.py [hunk 6]
```diff
@@ -144,6 +153,17 @@ def run(ceph_cluster, **kw):
         LOG.error("Failed during deployment")
         return rc
 
+    # Check logs has generated after running the deploy playbook
+    if custom_log:
```
Comment (psathyan) [inline tests/ceph_installer/test_ansible.py:157]: Shouldn't we also pass the custom log option to the ansible playbook?

File: suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml (hunk 3 above)
Comment (adityaramteke) [reply suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml:29]: @psathyan, previously, I added another test and it looked duplicate to this, hence @sunilangadi2 has suggested adding this custom_log config parameter to this test case and comment the polarion ID. To apply your suggestion, we'll need to add the two polarion ids in the polarion-id key!

File: tests/ceph_installer/test_ansible.py (hunk 5 above)
Comment (adityaramteke) [reply tests/ceph_installer/test_ansible.py:125]: Sure, will update the changes.

File: tests/ceph_installer/test_ansible.py (hunk 4 above)
Comment (adityaramteke) [reply tests/ceph_installer/test_ansible.py:130]: @psathyan, I was thinking the same previously! But, `ls` looks more simple to me and gives more options if we'll need to modify it in future. Let me know if you still want me to use `du`.

File: tests/ceph_installer/test_ansible.py (hunk 6 above)
Comment (adityaramteke) [reply tests/ceph_installer/test_ansible.py:157]: @psathyan, sorry I did not get you! Can you please explain it?

File: suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml (hunk 3 above)
Comment (psathyan) [reply suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml:29]: Automation requires that we should be able to pinpoint the test case that has failed. I believe this requirement is not met in this case hence the comment. Please discuss this in the FG

File: tests/ceph_installer/test_ansible.py (hunk 4 above)
Comment (psathyan) [reply tests/ceph_installer/test_ansible.py:130]: Here the requirement is to check the size of the file (existing) hence would prefer using `du` as it is straight forward.

File: tests/ceph_installer/test_ansible.py (hunk 6 above)
Comment (psathyan) [reply tests/ceph_installer/test_ansible.py:157]: Sorry for the confusion...

By default the `log_path` is set to `None`, hence shouldn't we modify the ansible.cfg variable `DEFAULT_LOG_PATH` when `custom_log` is set to True?

File: suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml (hunk 3 above)
Comment (adityaramteke) [reply suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml:29]: @psathyan  Thanks, I'll discuss it more within FG. To get more information on it! Can we add two or more polarion IDs to `polarion-id`?

File: tests/ceph_installer/test_ansible.py (hunk 4 above)
Comment (adityaramteke) [reply tests/ceph_installer/test_ansible.py:130]: Sure, will use `du` instead.

File: tests/ceph_installer/test_ansible.py (hunk 6 above)
Comment (adityaramteke) [reply tests/ceph_installer/test_ansible.py:157]: > Sorry for the confusion...
> 
> By default the `log_path` is set to `None`, hence shouldn't we modify the ansible.cfg variable `DEFAULT_LOG_PATH` when `custom_log` is set to True?

@psathyan, I believe that's not true, the default `log_path` will never be set to 'None'! Also, the test case is not meant to modify the default `log_path` but to look into the `log_path` and check if the logs are available. 

Let me know if I'm still missing something here! 

As per my knowledge the default `log_path` will be set to: 
``` log_path = $HOME/ansible/ansible.log ```

File: suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml (hunk 3 above)
Comment (psathyan) [reply suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml:29]: If we add more than one Polarion ID... we would not be able to pin point which test case has failed.

File: tests/ceph_installer/test_ansible.py (hunk 6 above)
Comment (psathyan) [reply tests/ceph_installer/test_ansible.py:157]: I don't have a 4.x setup to verify but going by https://docs.ansible.com/ansible/latest/reference_appendices/logging.html it is set to None.

Comment (adityaramteke) [reply tests/ceph_installer/test_ansible.py:157]: @psathyan, we are discussing the `ceph-ansible`! Our package is designed with the default configuration when you'll install the ceph-ansible rpm package `ansible.cfg` will get created at `/usr/share/ceph-ansible` with the default log location `$HOME/ansible/ansible.log` until and unless you change it to other location.

I hope this helps! :-)

File: suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml (hunk 3 above)
Comment (adityaramteke) [reply suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml:29]: @psathyan, Thanks, I get your point and I quickly discussed this within the FG, as per the discussion, we do not want to create a duplication of the test and hence need to update the polarion ID within this test! Please help me if you have any other suggestions for the same.

File: tests/ceph_installer/test_ansible.py (hunk 6 above)
Comment (psathyan) [reply tests/ceph_installer/test_ansible.py:157]: Thanks for clarifying it.

File: tests/ceph_installer/test_ansible.py (hunk 5 above)
Comment (adityaramteke) [reply tests/ceph_installer/test_ansible.py]: Updated the changes.

File: tests/ceph_installer/test_ansible.py (hunk 4 above)
Comment (adityaramteke) [reply tests/ceph_installer/test_ansible.py]: Updated the changes.

File: suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml (hunk 2 above)
Comment (sunilangadi2) [reply suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml]: Thanks @adityaramteke changes LGTM.

**Review by sunilangadi2 — APPROVED** [review]

**Comment by Ameenasuhani** [comment]
@adityaramteke Can you please fix the tox issue.

**Comment by adityaramteke** [comment]
> @adityaramteke Can you please fix the tox issue?

The issue has been fixed.

File: tests/ceph_installer/test_ansible.py [hunk 7]
```diff
@@ -169,6 +186,26 @@ def run(ceph_cluster, **kw):
...
```
Comment (vamahaja) [inline tests/ceph_installer/test_ansible.py:190]: @adityaramteke as `custom_log: true`, this condition will get satisfied and no code will be executed aftet this if block. It will affect current modified test as well. Please confirm if this is intended.

Comment (adityaramteke) [reply tests/ceph_installer/test_ansible.py:190]: @vamahaja, yes as per the test if the condition is not satisfied after deployment no followed code will get executed and it's intended.

File: suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml (hunk 3 above)
Comment (sunilangadi2) [reply suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml:29]: @psathyan: @manasagowri suggested to edit the existing test case `CEPH-83571467`  itself along with log validation step, even I like that idea, otherwise for polarion id purpose if we are going to create new test suite then it's increases in the pipeline execution time with all these new suites.
That is ok right @psathyan ?
what's your thoughts?

Comment (psathyan) [reply suites/nautilus/ansible/tier-2_deploy_test-ceph-rpms.yaml:29]: @sunilangadi2 yes, it is one of the options we can take.

File: tests/ceph_installer/test_ansible.py [hunk 8]
```diff
@@ -146,6 +147,22 @@ def run(ceph_cluster, **kw):
 
     LOG.info("Ceph ansible version " + ceph_installer.get_installed_ceph_versions())
 
+    # Check if ansilbe.log file has any logs initially
+    if custom_log:
+        ansible_log_file, _ = ceph_installer.exec_command(
+            cmd=f"cat {ansible_dir}/ansible.cfg | grep log_path | cut -d \" \" -f3 | tr -d '\n\r'".format(
```
Comment (psathyan) [inline tests/ceph_installer/test_ansible.py:153]: This would be
```
cmd=f"grep log_path {ansible_dir}/ansible.cfg | cut -d ' ' | tr -d '\n\r'"
```

File: tests/ceph_installer/test_ansible.py [hunk 9]
```diff
@@ -146,6 +147,22 @@ def run(ceph_cluster, **kw):
...
+                ansible_dir=ansible_dir
+            )
+        )
+        file_size_before, err = ceph_installer.exec_command(
+            cmd=f"du \"{ansible_log_file}\" | cut -f1 | tr -d '\n\r'".format(
```
Comment (psathyan) [inline tests/ceph_installer/test_ansible.py:158]: When using `f` strings, the right approach would be 

`cmd=f"du {ansible_log_file} | cut -f1 | tr -d '\n\r'"`

File: tests/ceph_installer/test_ansible.py [hunk 10]
```diff
@@ -169,6 +186,26 @@ def run(ceph_cluster, **kw):
         LOG.error("Failed during deployment")
         return rc
 
+    # Check logs has generated after running the deploy playbook
+    if custom_log:
+        file_size_after, err = ceph_installer.exec_command(
+            cmd=f"du \"{ansible_log_file}\" | cut -f1 | tr -d '\n\r'".format(
```
Comment (psathyan) [inline tests/ceph_installer/test_ansible.py:192]: This is incorrect. Kindly modify as per the above example.

File: tests/ceph_installer/test_ansible.py [hunk 11]
```diff
@@ -169,6 +186,26 @@ def run(ceph_cluster, **kw):
...
+                ansible_log_file=ansible_log_file
+            )
+        )
+
+        if "No such file or directory" in err:
+            LOG.error(
+                f"{ansible_log_file} file not generated after deployment".format(
```
Comment (psathyan) [inline tests/ceph_installer/test_ansible.py:199]: Likewise, please use the correct syntax

**Review by psathyan — COMMENTED** [review]
Please provide the test run logs using the latest code modifications.

File: tests/ceph_installer/test_ansible.py (hunk 8 above)
Comment (adityaramteke) [reply tests/ceph_installer/test_ansible.py:153]: @psathyan, thanks, I got confused last time! I now corrected the changes! Will test them and post them!

File: tests/ceph_installer/test_ansible.py (hunk 9 above)
Comment (adityaramteke) [reply tests/ceph_installer/test_ansible.py]: Updated all the changes.

**Comment by adityaramteke** [comment]
@psathyan, I've updated the logs with the latest code modification!

**Comment by adityaramteke** [comment]
I've rebased my changes!

**Review by sunilangadi2 — APPROVED** [review]

**Review by vamahaja — APPROVED** [review]
LGTM

File: tests/ceph_installer/test_ansible.py [hunk 12]
```diff
@@ -146,6 +147,18 @@ def run(ceph_cluster, **kw):
 
     LOG.info("Ceph ansible version " + ceph_installer.get_installed_ceph_versions())
 
+    # Check if ansilbe.log file has any logs initially
+    if custom_log:
+        ansible_log_file, _ = ceph_installer.exec_command(
+            cmd=f"grep log_path {ansible_dir}/ansible.cfg | cut -d ' ' | tr -d '\n\r'"
```
Comment (psathyan) [inline tests/ceph_installer/test_ansible.py:153]: I believe we would need to escape the backlash... this is the output from the log
```
2022-08-26 08:49:56,514 - INFO - cephci.test_ansible:110 - Running command du  | cut -f1 | tr -d '

' on 10.0.208.48 timeout 600
```

Comment (adityaramteke) [reply tests/ceph_installer/test_ansible.py:153]: @psathyan, we'll need a backslash as the value it is returning is with a new line, hence I've used `tr -d '\n\r'`

**Review by pranavprakash20 — APPROVED** [review]
LGTM

File: tests/ceph_installer/test_ansible.py (hunk 12 above)
Comment (pranavprakash20) [reply tests/ceph_installer/test_ansible.py:153]: @adityaramteke As @psathyan mentioned, the backslash has to be escaped. 

Please check the output of the sample code without escaping the backslash

cmd = f"du {ansible_log_file} | cut -f1 | tr -d '\n\r'" 
=> du pranav/temp | cut -f1 | tr -d '
'

where as with backslash escaped, 
cmd = f"du {ansible_log_file} | cut -f1 | tr -d '\\n\\r'"
=> du pranav/temp | cut -f1 | tr -d '\n\r'

which is the expected command, if I'm not wrong.

**Review by pranavprakash20 — CHANGES_REQUESTED** [review]
Please have a look at the changes suggested

File: tests/ceph_installer/test_ansible.py (hunk 12 above)
Comment (adityaramteke) [reply tests/ceph_installer/test_ansible.py:153]: @pranavprakash20, I'm still getting the same output for both with and without escaping the backslash! 

``` 
cmd = f"du {ansible_log_file} | cut -f1 | tr -d '\n\r'"
du aramteke/temp | cut -f1 | tr -d '
'

cmd = f"du {ansible_log_file} | cut -f1 | tr -d '\n\r'"
du aramteke/temp | cut -f1 | tr -d '
'
```
What's the actual difference?

Comment (adityaramteke) [reply tests/ceph_installer/test_ansible.py:153]: Got the requested changes now, 

```
cmd = f"du {ansible_log_file} | cut -f1 | tr -d '\\n\\r'"
```
Thanks, @pranavprakash20 and @psathyan! 
I'm updating the changes and will post the script result!

**Review by vamahaja — APPROVED** [review]

**Review by pranavprakash20 — APPROVED** [review]
LGTM
